design-proposals/klip-36-grace-stream-stream-joins.md

```diff
@@ -0,0 +1,102 @@
+# KLIP 36 - GRACE period for stream-stream joins
+
+**Author**: agavra | 
+**Release Target**: 0.14 | 
+**Status**: In Discussion | 
+**Discussion**: TBD
+
+**tl;dr:** _Support controlling the GRACE period for a stream-stream join to improve disk space
+utilization for small windows_
+           
+## Motivation and background
+
+As described in https://github.com/confluentinc/ksql/issues/6152, the ability to specify
+the grace period for stream-stream joins allows more control over the disk space utilization.
+This is because a window needs to stay on disk until the grace period has elapsed in order to
+ensure that any events that come in timestamped during an old window can be processed as long
+as the grace period is active. This means that for windows dramatically smaller than the grace
+period (e.g. a minute) we need to keep that level of granularity for the default 24 hour window.
+
+## What is in scope
+
+- Syntax for specifying grace period within stream-stream joins
+
+## What is not in scope
+
+- N/A
+
+## Value/Return
+
+Enables onboarding use cases that perform stream-stream joins on high cardinality, high throughput
+datasets with small windows without exhausting too much disk space.
+
+## Public APIS
+
+Stream-stream joins require use of `WITHIN` clauses:
+
+```
+JOIN <stream> WITHIN <time unit> ON <condition>
+```
+
+This will be expanded to support a more complex `withinExpression` that leverages the same
+syntax as aggregation windows:
+
+```
+JOIN <stream> WITHIN (SIZE <time unit>, GRACE PERIOD <time unit>) ON <condition>
+```
+
+The old syntax will still be supported for backwards compatibility.
+
+## Design
+
+The `WITHIN` clause will be converted into a `KsqlWindowExpression`, and then converted into
+a `JoinWindows grace(final Duration afterWindowEnd)` call on the `JoinWindows` in the 
+`StreamStreamJoinBuilder`.
+
+## Test plan
+
+We will add the usual QTT tests to ensure that the system respects the new retention limits
+and manually test to ensure that we clean up the RocksDB state for expired windows.
```

Comment:
It's too bad there's no easy way to automate this. I was wondering if we could do something similar to KsMaterializationFunctionalTest which verifies retention for windowed aggregates ([link](https://github.com/confluentinc/ksql/blob/4b81912b01a16a663abf5ee29d9583cf514b3635/ksqldb-engine/src/test/java/io/confluent/ksql/materialization/ks/KsMaterializationFunctionalTest.java#L652)) but we'd have to expose the state stores for that which feels like overkill for purposes of testing.

Reply:
I think the saving grace here is that we can test that we properly set the grace period, and then trust that Kafka Streams tests their public APIs (we can't test everything in Kafka Streams - but that's OK)